xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-sources/src/main/java/org/xwiki/notifications/sources/NotificationManager.java

```diff
@@ -108,13 +123,37 @@ List<CompositeEvent> getEvents(String userId, NotificationFormat format, int exp
      * @throws NotificationException if error happens
      * @since 12.6.1
      * @since 12.7RC1
+     * @deprecated Prefer using {@link #getRawEvents(String, NotificationFormat, int, Date, boolean, Date, List)} and
+     *             rely on {@link org.xwiki.notifications.GroupingEventStrategy} to get the composite events.
      */
+    @Deprecated(since = "15.4RC1")
     default List<CompositeEvent> getEvents(String userId, NotificationFormat format, int expectedCount, Date untilDate,
         boolean untilDateIncluded, Date fromDate, List<String> blackList) throws NotificationException
     {
         return getEvents(userId, format, expectedCount, untilDate, fromDate, blackList);
     }
 
+    /**
+     * Return events to display as notifications concerning the specified user.
+     *
+     * @param userId id of the user
+     * @param format format of the notifications
+     * @param expectedCount the maximum events to return
+     * @param untilDate do not return events happened after this date
+     * @param untilDateIncluded true if the passed {@code untilDate} should be included
+     * @param fromDate do not return events happened before this date
+     * @param blackList list of ids of blacklisted events to not return (to not get already known events again)
+     * @return the matching events for the user, could be less than expectedCount but not more
+     * @throws NotificationException if error happens
+     * @since 15.4RC1
+     */
+    @Unstable
+    default List<Event> getRawEvents(String userId, NotificationFormat format, int expectedCount, Date untilDate,
```

Comment:
I'm wondering if it's not a good occasion to pass some kind of configuration instead of so many parameters, since it's a lot more future-proof (if we need more later).

Reply:
Using NotificationParameters maybe?

Reply:
I don't know it very well, but if it already carries the same information, then why not yes.

Reply:
Yes it does, I'm not a huge fan of it, because it doesn't really have accessor: the fields are public. I'm wondering if we shouldn't break it once for all to set accessors...

Reply:
> it doesn't really have accessor: the fields are public

Indeed, strange one.